Tramstation crossing signals won't turn red when it's on the other side of the station

lessthnthree opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
After my last PR for Tramstation crossing lights, for some reason the lights are changing to red on West when you send the tram from East to Central, which obviously shouldn't happen. This changes the red distance so that doesn't happen.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: LT3
fix: Tramstation's crossing signals are actually fixed this time. Honest.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

lessthnthree committed:
Updates timing for red distance

Come on lights, why you gotta be like that.

Mothblocks reviewed (commented):
Where does this number come from? It would be significantly more preferable for this to be dynamic so that when changes are made to the map this bug doesn't come back.

lessthnthree commented:
I was VVing numbers on signals in round until it stopped doing it. It comes from calculating the distance of the tram from the next station.

`	// Everything will be based on position and travel direction
	var/signal_pos
	var/tram_pos
	var/tram_velocity_sign // 1 for positive axis movement, -1 for negative
	// Try to be agnostic about N-S vs E-W movement
	if(tram.travel_direction & (NORTH|SOUTH))
		signal_pos = y
		tram_pos = tram_part.y
		tram_velocity_sign = tram.travel_direction & NORTH ? 1 : -1
	else
		signal_pos = x
		tram_pos = tram_part.x
		tram_velocity_sign = tram.travel_direction & EAST ? 1 : -1

	// How far away are we? negative if already passed.
	var/approach_distance = tram_velocity_sign * (signal_pos - tram_pos)

	// Check for stopped state.
	// Will kill the process since tram starting up will restart process.
	if(!tram.travelling)
		// If super close, show red anyway since tram could suddenly start moving. If the tram could be approaching, show amber.
		if(abs(approach_distance) < red_distance_threshold)
			set_signal_state(XING_STATE_RED)
			return PROCESS_KILL
		if(abs(approach_distance) < amber_distance_threshold)
			set_signal_state(XING_STATE_AMBER)
			return PROCESS_KILL
		set_signal_state(XING_STATE_GREEN)
		return PROCESS_KILL

	// Check if tram is driving away from us.
	if(approach_distance < 0)
		// driving away. Green. In fact, in order to reverse, it'll have to stop, so let's go ahead and kill.
		set_signal_state(XING_STATE_GREEN)
		return PROCESS_KILL

	// OK so finally the interesting part where it's ACTUALLY approaching
	if(approach_distance <= red_distance_threshold)
		set_signal_state(XING_STATE_RED)
		return
	if(approach_distance <= amber_distance_threshold)
		set_signal_state(XING_STATE_AMBER)
		return
	set_signal_state(XING_STATE_GREEN)`

For Tramstation as it is, 35 causes the lights to turn red, and 33 it's fixed. Since it's already calculated based on distance between tram and signal, I don't really see how you'd get around that.

Mothblocks commented:
It's not calculated on distance if you had to manually change 25 to 33!

lessthnthree commented:
> It's not calculated on distance if you had to manually change 25 to 33!

Going 35 to 33. 35 seems just too close to the other station, so it turns it red before the proc runs again to update the lights. 

I think it's the if(!tram.travelling) part causing them to go red, but I don't know why that check is coming out with a different number, my guess is that the proc is only called every 2 seconds and the tram can basically go between an entire sector in that time.

Mothblocks commented:
My point is that ideally, you wouldn't have any numbers at all. What is going to happen if Tram is shrunk down or raised in size? How will another coder know when to adjust this number, and to what?

lessthnthree commented:
So right now it's doing X tiles from the crossing signal, and yeah I agree with you the number is arbitrary. But if you expanded Tram 5 tiles to the west, the crossing signal would indeed reflect that in when it changes. It'd be a complete rework with destination based signals, though that's ideally what I'd like to see at the end of the day.

lessthnthree committed:
Adds more info about the timing

Specific to Tramstation, but there is the possibility we have other maps in the future.

lessthnthree commented:
As present, the crossing signal checks the tram's distance and 35 happened to be a little too far away.

Mothblocks commented on code/modules/industrial_lift/crossing_signal.dm:
```diff
@@ -29,10 +29,10 @@
 	var/amber_distance_threshold = 60
-	/// Proximity threshold for red warning (running people will likely not be able to cross)
-	var/red_distance_threshold = 35
+	/// Proximity threshold for red warning (running people will likely not be able to cross) This is specific to Tramstation and may need to be adjusted if the map changes in the distance between tram stops.
```
I want more documentation on *how* you arrived to this number, so that when Tramstation is changed, we at least have a clue on how this number was reached.

lessthnthree replied:
I adjusted it 1 at a time until the signals were showing properly.

Mothblocks replied:
Document this as a comment, and describe what "showing properly" means

lessthnthree committed:
More info on signal timing

lessthnthree committed:
What are linebreaks?

Mothblocks commented on code/modules/industrial_lift/crossing_signal.dm:
```diff
@@ -29,10 +29,13 @@
 	/// Weakref to the tram piece we control
 	var/datum/weakref/tram_ref
-	/// Proximity threshold for amber warning (slow people may be in danger)
+	/// Proximity threshold for amber warning (slow people may be in danger) This is specific to Tramstation and may need to be adjusted if the map changes in the distance between tram stops.
```
```suggestion
	/// Proximity threshold for amber warning (slow people may be in danger).
	/// This is specific to Tramstation and may need to be adjusted if the map changes in the distance between tram stops.
```

Mothblocks committed:
Nit-make comment easier to read

Mothblocks commented:
We should still find a way to automate this better but this'll do

LemonInTheDark commented:
> My point is that ideally, you wouldn't have any numbers at all. What is going to happen if Tram is shrunk down or raised in size? How will another coder know when to adjust this number, and to what?

I agree this smells wrong
